### PR TITLE
adjoint/transpose transcend material concerns

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request is the next step in the JuliaLang/LinearAlgebra.jl#57 series after JuliaLang/julia#24969, JuliaLang/julia#25083, JuliaLang/julia#25125, JuliaLang/julia#25148, and JuliaLang/julia#25217. Specifically, this pull request makes `adjoint`/`transpose` uniformly lazy and extends ~`collect`~ `copy` for materialization.

Remaining tasks, for this or a downstream pull request:
1. ~Transform most `Adjoint`/`Transpose` calls to `adjoint`/`transpose`.~ done here
3. ~Lower `'` directly to `adjoint` (thereby fixing method overwrites during build).~ done here
2. Make `Adjoint`/`Transpose` behave like proper constructors.

Let's see what CI says. Thanks and best!

### CODE REVIEW

File: base/linalg/bitarray.jl [hunk 1]
```diff
@@ -123,7 +123,7 @@ end
 
 ## Structure query functions
 
-issymmetric(A::BitMatrix) = size(A, 1)==size(A, 2) && count(!iszero, A - transpose(A))==0
+issymmetric(A::BitMatrix) = size(A, 1)==size(A, 2) && count(!iszero, A - collect(A'))==0
```
Comment (mbauman) [inline base/linalg/bitarray.jl]: Just curious: is your rationale for `collect`ing here to ensure we still hit an optimized `-(::BitMatrix, ::BitMatrix)` method?  If that's the case it may be worth keeping note of this so we can revisit this PR for more `::Adjoint` method specializations in the future.

File: base/linalg/generic.jl [hunk 2]
```diff
@@ -986,7 +986,7 @@ function ishermitian(A::AbstractMatrix)
         return false
     end
     for i = indsn, j = i:last(indsn)
-        if A[i,j] != adjoint(A[j,i])
+        if A[i,j] != Adjoint(A[j,i])
```
Comment (mbauman) [inline base/linalg/generic.jl]: Why are you favoring the uppercase constructors in these cases?

File: base/number.jl [hunk 3]
```diff
@@ -67,9 +67,12 @@ function getindex(x::Number, I::Integer...)
     @boundscheck all([i == 1 for i in I]) || throw(BoundsError())
     x
 end
+# some code treats numbers as collection-like
 first(x::Number) = x
 last(x::Number) = x
-copy(x::Number) = x  # some code treats numbers as collection-like
+copy(x::Number) = x
+collect(x::Number) = x
```
Comment (mbauman) [inline base/number.jl]: This may be worth calling out and thinking through a bit more.  I expected `collect(1)` to return `[1]` since numbers are iterable "collections". In fact, right now `collect(1) = fill(1)` (that is, a zero-d array).

 I get why this change is needed, but I'm not entirely sold that it's the right answer.

File: base/linalg/bitarray.jl (hunk 1 above)
Comment (Sacha0) [reply base/linalg/bitarray.jl]: Partly, yes :). During the relevant transformation pass, preserving behavior was my primary concern. And for some array types, the rewrite above was strictly necessary to preserve behavior. (For example, `S + Adjoint(S)` for sparse matrix `S` hits an `AbstractArray` `+` fallback yielding an `Array` rather than the former `S + adjoint(S)`'s `SparseMatrixCSC`. Hence the transformation `S + adjoint(S)` -> `S + collect(S')` rather than `S + S'`, though `S + S'` should work eventually.) And indeed, after the gross functionality is complete quite a bit of cleanup will be in order :). https://github.com/JuliaLang/julia/issues/25331 is my present breadcrumb for this particular issue.

File: base/linalg/generic.jl (hunk 2 above)
Comment (Sacha0) [reply base/linalg/generic.jl]: The simple answer is that this pull request presently represents an intermediate state which will change substantially on next push. Specifically, most uppercase calls should then disappear :).

File: base/number.jl (hunk 3 above)
Comment (Sacha0) [reply base/number.jl]: Agreed, I'm not entirely sold on it either :).

**Comment by Sacha0** [comment]
(Marking triage for `collect` discussion.)

File: base/number.jl (hunk 3 above)
Comment (JeffBezanson) [reply base/number.jl]: I'm also -1 on this definition.

**Comment by vtjnash** [comment]
From discussion on triage, there was some consensus to go with the following:
- deprecate `collect` name:
  - use `Array` constructor instead in some cases (usually for realizing lazy operations like iterators)
  - use `copy` in other cases (probably most of these)
    - define `copy(x::AbstractArray) = Array(x)` (or some equivalent like `copyto!(similar(Array, x), x)`)
    - encourage specializations of `copy` when there's a better definition which can fold away extra wrappers or, failing that, can return an object of the same type

**Comment by Sacha0** [comment]
Rebased and completed rewrites to `adjoint`/`transpose` in base, test, and stdlib. Passing sparse, linalg, and stdlib test suites locally. So this pull request should be in shape after a sweep to replace `collect` methods/calls with `copy` as per triage. Best!

**Comment by Sacha0** [comment]
Rebased, squashed out all intermediate commits in the adjoint/transpose lazification process, and added a commit transforming collect to copy as per triage. The last commit is #25418, which I will drop once #25418 merges. Passed all impacted test suites locally. Best!

**Comment by Sacha0** [comment]
Added a commit lowering `'` directly to `adjoint` and further cleaned up the commit history; the lowering change incidentally fixes the method overwrite warning during build mentioned in https://github.com/JuliaLang/julia/pull/25212. I plan to make the `Adjoint`/`Transpose` constructor changes in a separate pull request. So this pull request should now be in shape. Best!

File: base/pkg/resolve/maxsum.jl [hunk 4]
```diff
@@ -122,7 +122,7 @@ mutable struct Graph
                         adjdict[p0][p1] = j1
 
                         bm = trues(spp[p1], spp[p0])
-                        bmt = adjoint(bm)
+                        bmt = trues(spp[p0], spp[p1])
```
Comment (mbauman) [inline base/pkg/resolve/maxsum.jl]: Wow, your re-write here is _way_ better.  Bravo for not just zoning out (like I did when this change caught me off guard).

**Review by mbauman — APPROVED** [review]
LGTM!  Very nice work.

File: base/pkg/resolve/maxsum.jl (hunk 4 above)
Comment (Sacha0) [reply base/pkg/resolve/maxsum.jl]: One small step towards atonement for the horde of ghastly transformations I introduced in #24969 😉.

**Comment by Sacha0** [comment]
Rebased out the last commit with #25418 in. With Matt and CI approving, absent objections or requests for time I plan to merge these changes Sunday evening PT or later. Best! :)

**Comment by Sacha0** [comment]
Rebased out conflict. Absent objections I plan to merge these changes once CI clears. Best!

**Comment by Sacha0** [comment]
(AV and FreeBSD failures are timeouts without apparent issue prior.)

**Comment by Sacha0** [comment]
Thanks all! :)
